pkg/objectcache/applicationprofilecache/applicationprofilecache.go

```diff
@@ -55,91 +36,252 @@ type ContainerCallStackIndex struct {
-	ap.userManagedProfiles.Set(baseProfileUniqueName, appProfile)
+	for {
+		select {
+		case <-ticker.C:
+			apc.updateAllProfiles(ctx)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// updateAllProfiles fetches all application profiles from storage and updates the cache
+func (apc *ApplicationProfileCacheImpl) updateAllProfiles(ctx context.Context) {
+	// Process namespace by namespace to optimize LIST operations
+	apc.namespaceToContainers.Range(func(namespace string, containerSet mapset.Set[string]) bool {
+		// Skip empty namespaces
+		if containerSet.Cardinality() == 0 {
+			return true
+		}
 
-	// If we have the base profile cached, fetch a fresh copy and merge.
-	// If the base profile is not cached yet, the merge will be attempted when it's added.
-	if ap.slugToAppProfile.Has(baseProfileUniqueName) {
-		// Fetch fresh base profile from cluster
-		freshBaseProfile, err := ap.getApplicationProfile(appProfile.GetNamespace(), baseProfileName)
+		// Get profiles list for this namespace
+		profileList, err := apc.storageClient.ApplicationProfiles(namespace).List(ctx, metav1.ListOptions{})
 		if err != nil {
-			logger.L().Warning("ApplicationProfileCacheImpl - failed to get fresh base profile for merging",
-				helpers.String("name", baseProfileName),
-				helpers.String("namespace", appProfile.GetNamespace()),
+			logger.L().Error("failed to list application profiles",
+				helpers.String("namespace", namespace),
 				helpers.Error(err))
-			return
+			return true // Continue to next namespace
+		}
+
+		// Process each profile
+		for _, profile := range profileList.Items {
+			// Handle user-managed profiles
+			if isUserManagedProfile(&profile) {
+				apc.handleUserManagedProfile(&profile)
+				continue
+			}
+
+			// Only consider completed profiles
+			if profile.Annotations[helpersv1.StatusMetadataKey] != helpersv1.Completed {
+				continue
+			}
+
+			// Get the workload ID from profile
+			workloadID := profile.Annotations[helpersv1.WlidMetadataKey]
+			if workloadID == "" {
+				continue
+			}
+
+			// Check if this workload ID is used by any container in this namespace
+			workloadIDInUse := false
+			for containerID := range containerSet.Iter() {
+				if containerInfo, exists := apc.containerIDToInfo.Load(containerID); exists &&
+					containerInfo.WorkloadID == workloadID {
+					workloadIDInUse = true
+					break
+				}
+			}
+
+			if !workloadIDInUse {
+				continue
+			}
+
+			// Update the profile in the cache
+			if existingProfile, exists := apc.workloadIDToProfile.Load(workloadID); exists {
+				// If the profile already exists and it's complete/completed, continue to the next one
+				if existingProfile.Annotations[helpersv1.CompletionMetadataKey] == helpersv1.Complete {
+					continue
+				}
+
+				// If the new profile is not complete and we already have a completed/partial one, skip it
+				if profile.Annotations[helpersv1.CompletionMetadataKey] != helpersv1.Complete {
+					continue
```

Comment:
I am confused here. We check at line 141 that we already have a completed profile, but we do not check if it was partial or complete. In theory, we prevent the update of any partially completed profile if we have a completed profile in the cache. Was this the purpose?

Reply:
Yes, if we hold in cache partial/completed I want to update if we get complete/completed. But if I have complete/completed I don't want to update.